### PR TITLE
📚 New theme & rearrangement of the docs

### PR DESCRIPTION
**Pull request by andersy005**
- [ ] Closes #xxxx
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by andersy005** [comment]
Cc @jhamman

**Comment by andersy005** [comment]
@pydata/xarray, does anyone know what's going on with readthedocs: https://readthedocs.org/projects/xray/builds/12928622/? I can't seem to figure out the error causing the build failures.

**Comment by keewis** [comment]
that's because of the headings in `whats-new.rst`. I think you have to be consistent with the style you use (`-`, `=`, `+`, `~`).

**Comment by DocOtak** [comment]
Also, the build step also has a `-W` flag which turns warnings into Errors and causes a non 0 exit status. This is probably because the read the docs config file for xarray has a `fail_on_warning` set to true

**Comment by andersy005** [comment]
> that's because of the headings in `whats-new.rst`. I think you have to be consistent with the style you use (`=`, `+`, `~`, `-`).


Ah okay... Is this issue connected to a specific sphinx version or readthedocs? I'm asking because I don't run into this issue locally

**Comment by andersy005** [comment]
> Also, the build step also has a `-W` flag which turns warnings into Errors and causes a non 0 exit status. This is probably because the read the docs config file for xarray has a `fail_on_warning` set to true

Thank you for the clarification....

**Comment by keewis** [comment]
> the build step also has a `-W` flag which turns warnings into Errors and causes a non 0 exit status

this is intentional: these errors usually cause the page layout to break somehow, so failing the sphinx build (especially in PRs) is important. We also try to minimize the number of nitpicky errors, but we don't enforce that since it seems almost impossible to silence all of them (we inherit docstrings from other projects like `numpy`, `pandas`, and `matplotlib`), and they only result in broken links.

**Comment by dcherian** [comment]
Anderson, this newest version looks really great.

I noticed that https://xray--4835.org.readthedocs.build/en/4835/index.html#see-also is really a list of tutorials and talks, which would be nice to highlight.

So I wonder if it makes sense to move "Getting Started" under "User Guide" and add a new top-level "Tutorials"

**Comment by andersy005** [comment]
> So I wonder if it makes sense to move "Getting Started" under "User Guide" and add a new top-level "Tutorials" 

@dcherian, I should point out that I created a new section [(`gallery`)](https://xray--4835.org.readthedocs.build/en/4835/gallery.html) and this is where I've been putting the examples/tutorials. Are you suggesting that we merge the `gallery`, `getting-started-guide` with the `user-guide`? Or should we just 
- merge the `getting-started` with the `user-guide` and 
- then move the https://xray--4835.org.readthedocs.build/en/4835/index.html#see-also to the gallery section?

**Comment by dcherian** [comment]
The tutorial/video resources are a bit hidden right now. It would be good to emphasize them. For e.g. matplotlib has (https://matplotlib.org/) 
```
Installation Documentation Examples Tutorials Contributing
```
in their top bar, so my suggestion was to add `Tutorials` in the top bar. That would make the top bar a bit crowded which is why I suggested merging `getting-started` with the `user-guide`.

Let's wait for some more opinions before making changes.

**Comment by andersy005** [comment]
Thank you for the pointer to the matplolib's doc structure. I hadn't figured out the distinction between examples and tutorials until I looked at their examples (which seem to focus on a single feature) and tutorials (which seem to be a complete workflow with deep-dives) pages. 

After looking at other projects, namely pymc3(https://docs.pymc.io/nb_tutorials/index.html, https://docs.pymc.io/nb_examples/index.html), scikit-learn (https://scikit-learn.org/stable/auto_examples/index.html, https://scikit-learn.org/stable/tutorial/index.html), I'm in favor of (1) having dedicated sections for both tutorials and examples and (2) merging the getting-started guide with user-guide.

**Comment by andersy005** [comment]
@keewis,

I recall you mentioning that the xarray's sphinx build on RTD is configured to turn warnings into errors. Currently, I have an extension that warns that it's not safe for parallel reads. As a result, the entire build fails:

> WARNING: the myst_nb extension is not safe for parallel reading
> WARNING: doing serial read

Can you suggest any workaround for this issue?

**Comment by keewis** [comment]
I don't know any, unfortunately. We do want to keep the warnings-as-errors because those tend to result in broken page layouts, but RTD insists on the parallel build (which *does* make the builds faster) and `myst-nb` doesn't seem to support it. Is there a particular reason why we're moving away from `nbsphinx`?

**Comment by keewis** [comment]
unrelated to that: the internals page is a collection of mostly unrelated topics. It might be good to break it up into separate pages (e.g. "Variable", "duckarray integration", "accessors", "zarr spec", "custom backends", etc.)

**Comment by andersy005** [comment]
> Is there a particular reason why we're moving away from `nbsphinx`?

I got a bit carried away when I saw the _Markdown_ design document in https://github.com/pydata/xarray/pull/4979 :). I figured with myst/myst-nb, it would be easier to support documentation source written in any format (`.md`, `.rst`, `.ipynb`) in the future. I will revert back to nbsphinx. 

> unrelated to that: the internals page is a collection of mostly unrelated topics. It might be good to break it up into separate pages (e.g. "Variable", "duckarray integration", "accessors", "zarr spec", "custom backends", etc.)

👍🏽. I am going to separate them into different pages

**Comment by andersy005** [comment]
@pydata/xarray, this is ready for another round of reviews...

File: doc/conf.py
Comment (keewis) [inline doc/conf.py:182]: we don't use `numpydoc` anymore, so this could probably be removed:
```suggestion
```

File: doc/conf.py
Comment (keewis) [inline doc/conf.py:83]: the edit links don't work anymore (they point at the generated files). I think this is because we switched themes and `scanpydoc.rtd_github_links` only works with `sphinx_rtd_theme` (not sure, though).

**Comment by dcherian** [comment]
Anderson, I like the RTD-like sidebar. I think every pydata user is conditioned to look at the left sidebar to navigate the documentation :) so preserving it is nice.

image

Can we add another level of organization to API reference in the sidebar: "Top-Level Functions", "Dataset" etc.?

Another minor comment: How about renaming "external resources" to "Tutorials & Videos"? Similarly "Related projects" to "Ecosystem"?

It'd be nice to get the left sidebar to "pin" like the right one. Right now if you scroll enough ,it disappears
image

In general, this is very awesome. I think we should merge soon and iterate further.

**Comment by keewis** [comment]
@andersy005, the copy button feature is a great find, copying examples is so much easier than before.

The only issue I have with the new theme is that due to the maximum line length we use some examples have horizontal scroll bars (unless the left side bar is hidden).

File: doc/conf.py
Comment (andersy005) [reply doc/conf.py]: After some digging on `scanpydoc` ext, it's my understanding that the approach used by `scanpydoc.rtd_github_links` is intertwined with the `sphinx_rtd_theme`. I was unable to get it to work with the new theme. To address this issue, I disabled the `edit on GitHub` button on the generated docstring pages, and used the `sphinx.ext.linkcode` extension to link the source code: 


Screen Shot 2021-03-06 at 7 17 00 PM

**Comment by andersy005** [comment]
@dcherian, 

> Can we add another level of organization to API reference in the sidebar: "Top-Level Functions", "Dataset" etc.?
 
> It'd be nice to get the left sidebar to "pin" like the right one. Right now if you scroll enough ,it disappears

I have yet to figure this out. If someone knows how to do this (whether with or without custom HTML template), please let me know :)

> Another minor comment: How about renaming "external resources" to "Tutorials & Videos"? Similarly "Related projects" to "Ecosystem"?

Done! 

@keewis, 

> The only issue I have with the new theme is that due to the maximum line length we use some examples have horizontal scroll bars (unless the left side bar is hidden).

I will see if there's a remedy for this.

File: doc/conf.py
Comment (andersy005) [inline doc/conf.py:291]: This is a list of pages that were moved to new locations. It'd be nice/convenient to set up redirects for these so as to avoid `404 errors`. I was going to use the `sphinx_reredirects` ext until I noticed that sphinx emits the following warning which results into build failures on ReadtheDocs:

```bash
WARNING: the sphinx_reredirects extension does not declare if it is safe for parallel reading, assuming it isn't - please ask the extension author to check and make it explicit
```


One remedy is to set these redirects from readthedocs settings page: https://docs.readthedocs.io/en/stable/user-defined-redirects.html

Comment (keewis) [reply doc/conf.py:291]: I made you a maintainer on RTD so you should be able to play around with that.

Did you try `sphinxext-rediraffe`? From a quick look this should do the same as `sphinx_reredirects`, but it seems more mature and does declare itself as safe for parallel read / write.

**Comment by keewis** [comment]
> I will see if there's a remedy for this.

You could try to reduce the (quite big) margin on the left / right

File: doc/conf.py
Comment (andersy005) [reply doc/conf.py]: > Did you try sphinxext-rediraffe? From a quick look this should do the same as sphinx_reredirects, but it seems more mature and does declare itself as safe for parallel read / write.

I just tried `sphinxext-rediraffe`, and it works like a charm. Thank you for the suggestion

**Comment by andersy005** [comment]
> It'd be nice to get the left sidebar to "pin" like the right one. Right now if you scroll enough ,it disappears 

@dcherian, this is fixed. When you get a moment, take another look and let me know if I missed anything.

**Comment by andersy005** [comment]
> The only issue I have with the new theme is that due to the maximum line length we use some examples have horizontal scroll bars (unless the left side bar is hidden).

@keewis, I fixed this issue. 


@pydata/xarray, this PR is ready.  Unless there's any additional issue that needs to be addressed before merging, feel free to merge this.

**Review by dcherian — APPROVED** [review]

**Review by keewis — APPROVED** [review]
> @keewis, I fixed this issue.

thanks, I can confirm this is fixed.

Looking at `pydata-sphinx-theme` and `sphinx-book-theme`, both seem to have a `doc_path` or `path_to_docs` context variable which allows specifying a custom path. Even if we get the edit-on-github feature to work I would keep the source link, though.

The reasoning for the current behavior of the edit-on-github feature was that people were opening PRs to be merged into `stable`, which would then be dropped on a release. We could avoid that by either always linking to `master` (but then we can't determine the line numbers; this is what we are doing with `scanpydoc`), or by always opening a PR which also adds the commit to `master` (using a bot?). Another option would be to restrict PRs into `stable` to maintainers, but I'm not sure if github supports that.

File: doc/conf.py
Comment (andersy005) [inline doc/conf.py:228]: > Looking at pydata-sphinx-theme and sphinx-book-theme, both seem to have a doc_path or path_to_docs context variable which allows specifying a custom path (I think we can merge this PR in its current state and fix this and any other issues in a new PR, though). 
> The reasoning for the current behavior of the edit-on-github feature was that people were opening PRs to be merged into stable, which would then be dropped on a release. We could avoid that by either always linking to master (but then we can't determine the line numbers; this is what we are doing with scanpydoc), or by always opening a PR which also adds the commit to master (using a bot?)

@keewis, unless I am misunderstanding how the existing `edit on Github` button works, I think these settings allow us to replicate the existing behavior i.e. For (1) regular pages, the button links to the document/file in the master branch. For (2) API pages, the button is disabled and the `[source] link` links to the lines of code in the master branch. Could you elaborate on when/how with the new changes, the `edit on GitHub` button will result in pointing people to the wrong branch?

Comment (keewis) [reply doc/conf.py:228]: the edit on github link will currently point to the same branch / tag / commit the docs were built for (regardless of the page).

The existing fix (always point to `master`) makes it impossible to accurately determine the line numbers for API pages. Since we might even have removed that function I'm starting to think that it might be easier to not redirect at all but rather make the fact that the PR does not point to `master` more visible.

Not sure what the best way to do so is, though. Maybe have a workflow post a comment if the branch is not the dev branch?

Comment (andersy005) [reply doc/conf.py:228]: Thank you for the clarification! 

> Not sure what the best way to do so is, though. Maybe have a workflow post a comment if the branch is not the dev branch?

I presume it'd be okay to merge this PR and then address this in a separate PR in the near future?

Comment (keewis) [reply doc/conf.py:228]: for sure!

**Comment by andersy005** [comment]
I plan to merge this later today (unless there's any objection and/or necessary changes)

**Comment by shoyer** [comment]
+1

**Comment by andersy005** [comment]
Thank you @keewis, @dcherian, @shoyer!

**Pull request merged**
